tests/Unit/ControlSystem/Systems/Test_RotScaleTrans.cpp

```diff
@@ -0,0 +1,240 @@
+    return binary_trajectories.positions(time);
+  };
+
+  // Run the actual control system test.
+  system_helper.run_control_system_test(runner, final_time, make_not_null(&gen),
+                                        position_function, coord_map);
+
+  // Grab results
+  const std::array<double, 3> grid_position_of_a =
+      system_helper.grid_position_of_a();
+  const std::array<double, 3> grid_position_of_b =
+      system_helper.grid_position_of_b();
+
+  // Our expected positions are just the initial positions
+  const std::array<double, 3> expected_grid_position_of_a{
+      {-0.5 * initial_separation, 0.0, 0.0}};
+  const std::array<double, 3> expected_grid_position_of_b{
+      {0.5 * initial_separation, 0.0, 0.0}};
+
+  const auto& rotation_f_of_t = dynamic_cast<
+      domain::FunctionsOfTime::QuaternionFunctionOfTime<RotationDerivOrder>&>(
+      *functions_of_time.at(rotation_name));
+  const auto& translation_f_of_t = dynamic_cast<
+      domain::FunctionsOfTime::PiecewisePolynomial<TranslationDerivOrder>&>(
+      *functions_of_time.at(translation_name));
+  const auto& expansion_f_of_t = dynamic_cast<
+      domain::FunctionsOfTime::PiecewisePolynomial<ExpansionDerivOrder>&>(
+      *functions_of_time.at(expansion_name));
+
+  const auto omega = rotation_f_of_t.angle_func_and_deriv(final_time)[1];
+  const auto trans_and_2_derivs =
+      translation_f_of_t.func_and_2_derivs(final_time);
+  const auto expansion_factor = expansion_f_of_t.func(final_time)[0][0];
+
+  // The control system gets more accurate the longer you run for. This is
+  // the accuracy we can achieve in this amount of time.
+  Approx custom_approx1 = Approx::custom().epsilon(5.0e-5).scale(1.0);
+  Approx custom_approx2 = Approx::custom().epsilon(rotation_eps).scale(1.0);
+
+  const DataVector expected_omega{
+      {0.0, 0.0, binary_trajectories.angular_velocity(final_time)}};
+  const DataVector expected_translation_2nd_deriv{3, 0.0};
+  CHECK_ITERABLE_CUSTOM_APPROX(expected_omega, omega, custom_approx1);
+  CHECK_ITERABLE_CUSTOM_APPROX(trans_and_2_derivs[2],
+                               expected_translation_2nd_deriv, custom_approx1);
+  CHECK_ITERABLE_CUSTOM_APPROX(trans_and_2_derivs[1],
+                               array_to_datavector(initial_velocity),
+                               custom_approx1);
+  CHECK_ITERABLE_CUSTOM_APPROX(
+      trans_and_2_derivs[0], array_to_datavector(initial_velocity * final_time),
+      custom_approx1);
+  CHECK_ITERABLE_CUSTOM_APPROX(
+      binary_trajectories.separation(final_time) / initial_separation,
+      expansion_factor, custom_approx1);
+
+  CHECK_ITERABLE_CUSTOM_APPROX(expected_grid_position_of_a, grid_position_of_a,
+                               custom_approx2);
+  CHECK_ITERABLE_CUSTOM_APPROX(expected_grid_position_of_b, grid_position_of_b,
+                               custom_approx2);
+}
+}  // namespace
+
+// [[Timeout, 15]]
+SPECTRE_TEST_CASE("Unit.ControlSystem.Systems.RotScaleTrans",
```

Comment:
Can this test be sped up?

Reply:
I don't see an obvious way other than removing some of the tests. When I test the individual control systems, I run the whole test twice (once for deriv order 2, once for deriv order 2) and that fits within the default 5 second timeout. However, here I'm testing all three at once with all combos of deriv orders which is 7 tests in total. So I'd expect this whole test to take ~3.5 times as long as the individual ones.

We probably won't ever use rotation with deriv order 2 as it's just not as accurate as deriv order 3 so I could potentially remove those tests to speed things up. What do you think?

Reply:
Knowing nothing about this, naively I'm a bit worried about how slow the test is simply because it could signal the control system calculations are much slower or less efficient than they should be. But I'll emphasize that I haven't thought about this at all and that just having a 15 second unit test tends to indicate something's off. For comparison, testing the time derivative action for conservative, nonconservative, and mixed systems, with and without a moving mesh, in 1-3d all fits in less than 2 seconds. Naively that seems like it would be a much more expensive test than a bunch of 1d fits

Reply:
Ok I know why it's so slow. It's the rotation part. It solves a whole ODE to get the quaternion every time the map is evaluated or updated.

Damping time 0.5 * update fraction 0.3 / measurements per update 4 = time between measurements 0.0375 (this is how often the map is evaluated in one test initially. It does settle eventually, but the max damp time is 10.0 which equates to 0.75 between measurements). Each test goes to time 500. So that's on the order of 10,000 map evaluations per test. And there are 7 different cases tested so that's ~70,000 map evaluations total...so it makes sense that this is sort of slow.

I removed rotation control from these tests and just had expansion and translation together. The test with rotation averaged 6.1 seconds locally. The test without rotation averaged 2.2 seconds locally.

This has come up before [here](https://github.com/sxs-collaboration/spectre/pull/3817#discussion_r813282989) and [here](https://github.com/sxs-collaboration/spectre/pull/3817#discussion_r825417918).

Reply:
Not sure what to do about this. I believe the fix is a partial redesign of how we deal with damping times and updating functions of time. I'm currently working this.